core/src/main/java/jenkins/model/GlobalBuildDiscarderListener.java

```diff
@@ -46,6 +46,15 @@ public class GlobalBuildDiscarderListener extends RunListener<Run> {
     @Override
     public void onFinalized(Run run) {
         Job job = run.getParent();
-        BackgroundGlobalBuildDiscarder.processJob(new LogTaskListener(LOGGER, Level.FINE), job);
+        try {
+            // Job-level build discarder execution is unconditional.
+            job.logRotate();
+        } catch (Exception e) {
+            LOGGER.log(Level.WARNING, e, () -> "Failed to rotate log for " + run);
+        }
+        // Avoid calling Job.logRotate twice in case JobGlobalBuildDiscarderStrategy is configured globally.
```

Comment:
This is a bit confusing, but essentially the old behavior was that `Run.run` called `Job.logRotate` while the build was in `POST_PRODUCTION` state, but then if the configuration of `GlobalBuildDiscarderConfiguration` included `JobGlobalBuildDiscarderStrategy` (default behavior), we also called `Job.logRotate` here while the build was in `COMPLETED` state. We only need it to be called once, and it's preferable to call it here so that we can check `Run.isLogUpdated` everywhere.

Also, this old double log rotation is why I am not concerned about removing [this call](https://github.com/jenkinsci/workflow-job-plugin/blob/d82e9f82c8115f04fa3f27fc38e18ca01c6410d2/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L663-L669) in `workflow-job` and also why I am not worried about making the above call asynchronous. `BackgroundGlobalBuildDiscarder.processJob` has been calling `logRotate` synchronously for years now in default configurations, so I think that the asynchronous behavior in https://github.com/jenkinsci/workflow-job-plugin/commit/63fdbe8c2d0f74f97e8f3b92cc566e2fff52bdf3 is no longer critical.

Reply:
(https://github.com/jenkinsci/workflow-job-plugin/pull/70 for better linking)